src/neurotechdevkit/imaging/beamform.py

```diff
@@ -0,0 +1,356 @@
+    LANZCOS_5 = 6  # 5-lobe Lanczos interpolation
+
+
+def beamform_delay_and_sum(
+    iq_signals: npt.NDArray[np.float_],
+    x: npt.NDArray[np.float_],
+    z: npt.NDArray[np.float_],
+    freq_sampling: float,
+    freq_carrier: float,
+    **kwargs,
+) -> npt.NDArray[np.float_]:
+    """Delay-And-Sum beamforming.
+
+    Currently, assumes that the input signal is in-phase/quadrature (I/Q) signals.
+
+    Parameters:
+        iq_signals: 2-D complex-valued array containing I/Q signals.
+            Shape: (time_samples, num_channels) or
+                (time_samples, num_channels, num_echoes)
+            channels usually correspond to transducer elements.
+        x: 2-D array specifying the x-coordinates of the [x, z] image grid.
+        z: 2-D array specifying the z-coordinates of the [x, z] image grid.
+        freq_sampling: sampling frequency of the input signals.
+        freq_carrier: center/carrier frequency of the input signals.
+        **kwargs: additional arguments passed to `delay_and_sum_matrix`
+
+    Returns:
+        Beamformed signals at the specified [x, z] image grid.
+
+    Notes:
+        - For a linear array: x = 0, z = 0 corresponds to the center of the transducer.
+            The x-axis is parallel to the transducer, and the z-axis is perpendicular
+            to the transducer, with z increasing in the direction of the transducer.
+        - Memory usage for the delay-and-sum matrix is:
+            O(num_pixels * num_channels * num_interpolation_points).
+
+    Adapted from: https://doi.org/10.1016/j.ultras.2020.106309
+    """
+    # Check input arguments
+    if iq_signals.ndim == 2:
+        num_time_samples, num_channels = iq_signals.shape
+    elif iq_signals.ndim == 3:
+        num_time_samples, num_channels, num_echoes = iq_signals.shape
+    else:
+        raise ValueError(
+            "Expected iq_signals to have shape (time_samples, num_channels) or "
+            "(time_samples, num_channels, num_echoes)."
+        )
+
+    assert np.iscomplexobj(
+        iq_signals
+    ), "Expected iq_signals to be complex-valued I/Q signals."
+    assert x.ndim == 2, "Expected x to have shape (width_pixels, depth_pixels)."
+    assert z.ndim == 2, "Expected z to have shape (width_pixels, depth_pixels)."
+    assert x.shape == z.shape, "Expected image grid x and z to have the same shape."
+
+    # Check for potential underlying errors before allocating memory
+    das_matrix: csr_array = delay_and_sum_matrix(
+        num_time_samples=num_time_samples,
+        num_channels=num_channels,
+        x=x,
+        z=z,
+        freq_sampling=freq_sampling,
+        freq_carrier=freq_carrier,
+        **kwargs,
+    )
+
+    # Beamform
+    iq_signals_column_vec = iq_signals.reshape(-1, 1)
+    beamformed_iq_signals = das_matrix @ iq_signals_column_vec
+    beamformed_iq_signals = beamformed_iq_signals.reshape(x.shape)
+
+    return beamformed_iq_signals
+
+
+def delay_and_sum_matrix(
```

Comment:
I wonder if splitting this function into various smaller parts performing smaller steps makes sense. Would it become more readable? or the way it's currently written is better?

Reply:
Updated!